File: 25-adv-ds2/1-intro/main.cpp
// Advanced Data Structures 2 - Intro C++ example
// Bit manipulation: Counting set bits

#include <iostream>

int countSetBits(int n) {
    int count = 0;
    while (n) {
        n &= (n - 1);
        count++;
    }
    return count;
}

int main() {
    int num = 29; // Binary: 11101
    std::cout << countSetBits(num) << std::endl;  // Output: 4
    return 0;
}

File: 25-adv-ds2/1-intro/main.java
// Advanced Data Structures 2 - Intro Java example
// Bit manipulation: Counting set bits

class Main {
    public static int countSetBits(int n) {
        int count = 0;
        while (n != 0) {
            n &= (n - 1);
            count++;
        }
        return count;
    }

    public static void main(String[] args) {
        int num = 29; // Binary: 11101
        System.out.println(countSetBits(num));  // Output: 4
    }
}
